docs/architecture/how/receipt-congestion.md

```diff
@@ -0,0 +1,211 @@
+Near Protocol executes transactions in multiple steps, or receipts. Once a
+transaction is accepted, the system has committed to finish all those receipts
+even if it does not know ahead of time how many receipts there will be or on
+which shards they will execute.
+
+This naturally leads to the problem that if shards just keep accepting more
+transactions, we might accept workload at a higher rate than we can execute.
+
+## Cross-shard congestion as flow problem
+
+For a quick formalized discussion on congestion, let us model the Near Protocol
+transaction execution as a [flow
+network](https://en.wikipedia.org/wiki/Flow_network).
+
+Each shard has a source that accepts new transactions and a sink for burning
+receipts. The flow is measured in gas. Edges to sinks have a capacity of 1000
+Tgas. (Technically, it should be 1300 but let's keep it simple for this
+discussion.)
+
+![graph](../../images/congestion/base_flow_network.svg)
+
+The edges between shards are not limited in this model. In reality, we are
+eventually limited by the receipt sizes and what we can send within a block time
+through the network links. But if we only look at that limit, we can send very
+many receipts with a lot of gas attached to them. Thus, the model considers it
+unlimited.
+
+Okay, we have the capacities of the network modeled. Now let's look how a
+receipt execution maps onto it.
+
+Let's say a receipt starts at shard 1 with 300 Tgas. While executing, it burns 100 Tgas and
+creates an outgoing receipts with 200 Tgas to another shard. We can represent this in the flow network with 
+100 Tgas to the sink of shard 1 and 200 Tgas to shard 2. 
+
+![graph](../../images/congestion/receipt_flow_example_0.svg)
+
+Note: The graph includes the execution of the next block with the 200 Tgas to th
+sink of shard 2. This should be interpreted as if we continue sending the exact
+same workload on all shards every block. Then we reach this steady state where
+we continue to have these gas assignments per edge.
+
+Now we can do som flow analysis. It is immediately obvious that the total
+outflow per is limited to N * 1000 Tgas but the incoming flow is unlimited.
+
+For a finite amount of time, we can accept more inflow than outflow, we just have to add buffers to store what we cannot execute, yet. But to stay within finite memory requirements, we need to fall back to a flow diagram where outflows are greater or equal to inflows within a finite time frame.
+
+Next, we look at a ideas one at the time before combining some of them into the
+cross-shard congestion design proposed in
+[NEP-539](https://github.com/near/NEPs/pull/539).
+
+## Idea 1: Compute the minimum max-flow and stay below that limit
+
+One approach to solve congestion would be to never allow more work into the
+system than we can execute.
+
+But this is not ideal Just consider this example where everybody tries to access
+a contract on the same shard.
+
+![graph](../../images/congestion/receipt_flow_example_1.svg)
```

Comment:
Depending on how it's interpreted the graph may also need a 999 Tgas arrow from shard 3 to itself and then the arrow to sink should be 2997 Tgas.

Reply:
Same for the graph below.

Reply:
I agree for the arrow shard3 -> shard3. But I think it should still be 2998 Tgas. Otherwise the sum of flows from the source is not equal to the sum to the sinks. It would be 2997 Tgas for the receipts and 1 Tgas for conversion, both should point to sink, showing that the gas has been burnt.

I'm adding the arrows to the two graphs as best as I can without it getting too unreadable... Aligning things isn't always easy in graphviz graphs^^